// Copyright 2020 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto2";

import "google/protobuf/timestamp.proto";

message Money {
  optional int32 dollars = 1;
  optional int32 cents = 2;

  enum Currency {
    UNKNOWN = 0;
    USD = 1;
    CAD = 2;
  }
  optional Currency currency = 3;
}

message Amount {
  oneof amount_demux {
    string str_amount = 1;
    int32 int_amount = 2;
    double double_amount = 3;
    google.protobuf.Timestamp timestamp_amount = 4;
    Money money_amount = 5;
    bool bool_amount = 6;
  }
}

message PointLocation {
  optional int32 row = 1;
  optional int32 col = 2;
}

// either cell-to-cell A1:B2
//     or cell-to-row  A1:3 (A1->A3)
//     or cell-to-col  A1;B (A1->B1)
//     or col-to-col   A:B
//     or row-to-row   2:3
message RangeLocation {
  oneof from_demux {
    PointLocation from_cell = 1;
    int32 from_row = 2;
    int32 from_col = 3;
  }

  oneof to_demux {
    PointLocation to_cell = 4;
    int32 to_row = 5;
    int32 to_col = 6;
  }
  // some of these states will be invalid.
}

message Expression {
  message Operation {
    optional string fn_name = 1;
    repeated Expression terms = 2;
  }

  oneof expression_demux {
    PointLocation lookup = 1;
    RangeLocation range = 2;
    Amount value = 3;
    Operation operation = 4;
  }
}

message Formula {
  optional Expression expression = 1;
  optional Amount cached_amount = 2;
}

message Cell {
  optional PointLocation point_location = 1;
  optional Formula formula = 2;
}

message Metadata {
  optional string title = 1;
  optional string author = 2;
  optional google.protobuf.Timestamp created_time = 3;
  optional google.protobuf.Timestamp edited_time = 4;
}

message LatisMsg {
  repeated Cell cells = 1;
  optional Metadata metadata = 2;
}
